clients/src/main/java/org/apache/kafka/common/utils/Utils.java

```diff
@@ -888,4 +888,26 @@ public static void writeTo(DataOutput out, ByteBuffer buffer, int length) throws
         return res;
     }
 
+    /**
+     * Stringify an exception.
+     *
+     * This includes the full stack trace-- we don't omit any frames.
+     * Every "cause" exception is also included.
+     *
+     * @param exception     The exception.
+     * @return              The exception as a string.
+     */
+    public final static String stringifyException(Throwable exception) {
+        StringWriter sw = new StringWriter();
+        String prefix = "";
+        for (Throwable e = exception; e != null; e = e.getCause()) {
+            sw.append(String.format("%s%s: %s\n", prefix,
+                e.getClass().getCanonicalName().toString(), e.getMessage()));
+            prefix = "Caused by: ";
+            for (StackTraceElement element : e.getStackTrace()) {
+                sw.append(String.format("        %s\n", element.toString()));
+            }
+        }
+        return sw.toString();
```

Comment:
There is a `stacktrace` method in this class. How is this different?

Reply:
`Utils#stackTrace` simply invokes `Throwable#printStackTrace` in order to generate a stack trace.  However, the stack trace generated this way can be incomplete.  Taking the example from the Java documentation pages, it may generate something like this:

```
 HighLevelException: MidLevelException: LowLevelException
         at Junk.a(Junk.java:13)
         at Junk.main(Junk.java:4)
 Caused by: MidLevelException: LowLevelException
         at Junk.c(Junk.java:23)
         at Junk.b(Junk.java:17)
         at Junk.a(Junk.java:11)
         ... 1 more
 Caused by: LowLevelException
         at Junk.e(Junk.java:30)
         at Junk.d(Junk.java:27)
         at Junk.c(Junk.java:21)
         ... 3 more
```

This is often frustrating because the important stack frame can get snipped and replaced with "10 more", etc.  So stringifyException simply iterates through all the stack frames and prints them.